src/core/metadata/qgslayermetadata.cpp

```diff
@@ -273,6 +278,460 @@ void QgsLayerMetadata::readFromLayer( const QgsMapLayer *layer )
+    keywordsElement.setAttribute( QStringLiteral( "vocabulary" ), i.key() );
+    const QStringList values = i.value();
+    for ( const QString &kw : values )
+    {
+      QDomElement keyword = document.createElement( QStringLiteral( "keyword" ) );
+      QDomText keywordText = document.createTextNode( kw );
+      keyword.appendChild( keywordText );
+      keywordsElement.appendChild( keyword );
+    }
+    metadataElement.appendChild( keywordsElement );
+  }
+
+  // fees
+  QDomElement fees = document.createElement( QStringLiteral( "fees" ) );
+  QDomText feesText = document.createTextNode( mFees );
+  fees.appendChild( feesText );
+  metadataElement.appendChild( fees );
+
+  // constraints
+  for ( const QgsLayerMetadata::Constraint &constraint : mConstraints )
+  {
+    QDomElement constraintElement = document.createElement( QStringLiteral( "constraints" ) );
+    constraintElement.setAttribute( QStringLiteral( "type" ), constraint.type );
+    QDomText constraintText = document.createTextNode( constraint.constraint );
+    constraintElement.appendChild( constraintText );
+    metadataElement.appendChild( constraintElement );
+  }
+
+  // rights
+  for ( const QString &right : mRights )
+  {
+    QDomElement rightElement = document.createElement( QStringLiteral( "rights" ) );
+    QDomText rightText = document.createTextNode( right );
+    rightElement.appendChild( rightText );
+    metadataElement.appendChild( rightElement );
+  }
+
+  // license
+  for ( const QString &license : mLicenses )
+  {
+    QDomElement licenseElement = document.createElement( QStringLiteral( "license" ) );
+    QDomText licenseText = document.createTextNode( license );
+    licenseElement.appendChild( licenseText );
+    metadataElement.appendChild( licenseElement );
+  }
+
+  // encoding
+  QDomElement encoding = document.createElement( QStringLiteral( "encoding" ) );
+  QDomText encodingText = document.createTextNode( mEncoding );
+  encoding.appendChild( encodingText );
+  metadataElement.appendChild( encoding );
+
+  // crs
+  QDomElement crsElement = document.createElement( QStringLiteral( "crs" ) );
+  mCrs.writeXml( crsElement, document );
+  metadataElement.appendChild( crsElement );
+
+  // extent
+  QDomElement extentElement = document.createElement( QStringLiteral( "extent" ) );
+
+  // spatial extents
+  const QList< QgsLayerMetadata::SpatialExtent > &sExtents = extent().spatialExtents();
+  for ( const QgsLayerMetadata::SpatialExtent &spatialExtent : sExtents )
+  {
+    QDomElement spatialElement = document.createElement( QStringLiteral( "spatial" ) );
+    // Dimensions fixed in the XSD
+    spatialElement.setAttribute( QStringLiteral( "dimensions" ), QStringLiteral( "2" ) );
+    spatialElement.setAttribute( QStringLiteral( "crs" ), spatialExtent.extentCrs.authid() );
+    spatialElement.setAttribute( QStringLiteral( "minx" ), spatialExtent.bounds.xMinimum() );
```

Comment:
better wrap the xMinimum(), etc in a qgsDoubleToString() (e.g. `qgsDoubleToString( spatialExtent.bounds.xMinimum()  )`. This will avoid any loss of precision when serializing/deserializing the value.